### PR TITLE
Add optionals 356

### LINKED ISSUE #356: Add Optionals
Where a null is used as a return type, change this to use an `Optional` instead. We don't want to use this for parameters

This is in line with a change to our style guide.

Branch: `0.2.0`

### PR DESCRIPTION
**Pull request by hieu-t-tran**
## Summary of change

Use `Optional` for all instances of `return null`. Update methods affected by this change for compatibility.

## Related issue

Closes #356

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [X] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [X] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [X] Ensure code changes pass all tests

- [X] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [X] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [X] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [X] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [X] Set this pull request to 'draft' if you are still working on it

- [X] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-jcijyenp-JiKFGBv62FIPoFnvOW6Ubg)

### CODE REVIEW

File: src/main/java/com/karankumar/bookproject/ui/book/form/BookForm.java [hunk 1]
```diff
@@ -377,32 +377,32 @@ private void populateBookShelf() {
 
     private void repopulateIfCreatingANewBook() {
         if (binder.getBean() == null) {
-            Book book = populateBookBean();
-            if (book != null) {
-                binder.setBean(book);
+            Optional<Book> book = populateBookBean();
+            if (book.isPresent()) {
+                binder.setBean(book.get());
             }
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/form/BookForm.java:383]: Could you please change this to 
```java
book.ifPresent(bean -> binder.setBean(bean));
```

File: src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java [hunk 2]
```diff
@@ -35,6 +35,6 @@
 
     @Override
     public Double convertToPresentation(RatingScale rating, ValueContext valueContext) {
-        return RatingScale.toDouble(rating);
+        return RatingScale.toDouble(rating).orElse(0.0);
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java:38]: In this case, we do want to return `null` in the else clause. Otherwise, this will change the semantics (`null` rating = no rating, which is different from a rating of 0/10).

We can change this to the following:

```java
return RatingScale.toDouble(rating).get();
```

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this! 

Could you also add optionals to the following places:

- `findByShelfNameAndUser` in `CustomShelfRepository.java`
- `findByPredefinedShelfNameAndUser` in `PredefinedShelfRepository.java`
- `findById()` in `AuthorService.java`
- `findById()` in `BookService.java`
- `findById()` in `CustomShelfService.java`
- `getCustomShelfByName()` in `CustomShelfService.java`
- `findById()` in `PredefinedShelfService.java`
- `findToReadShelf()` in `PredefinedShelfService.java`
- `findReadingShelf()` in `PredefinedShelfService.java`
- `findReadShelf()` in `PredefinedShelfService.java`
- `findDidNotFinishShelf()` in `PredefinedShelfService.java`
- `findByPredefinedShelfNameAndLoggedInUser()` in `PredefinedShelfService.java`
- `getPredefinedShelfByNameAsString()` in `PredefinedShelfService.java`
- `findById()` in `ReadingGoalService.java`
- `findById()` in `TagService.java`
- `findMostReadGenre()` in `GenreStatistics.java`
- `findMostLikedGenre()` in `GenreStatistics.java`
- `findLeastLikedGenre()` in `GenreStatistics.java`
- `findBookWithMostPages()` in PageStatistics.java
- `addSeriesToTitle()` in `BookUtils.java`

File: src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java (hunk 2 above)
Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java:38]: This came to my mind when I worked on this part as well. Thanks a lot for clarifying!

Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java:38]: It seems that just using `get()` here will produce some errors with `BookFormTest.java`. Using `orElse(null)` will get rid of those errors, but I'm not sure if this is the best approach since our goal is to reduce the usage of `null` in the first place.

Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java:38]: > It seems that just using get() here will produce some errors with BookFormTest.java.

That's strange, it shouldn't. Perhaps you're getting a Flyway error and need to clean? Either way, if you need a hand with this, feel free to ask on the #help channel on Slack and post the stack trace using something like GitHub Gist or Pastebin.

> but I'm not sure if this is the best approach since our goal is to reduce the usage of null in the first place.

Generally speaking, this is true to the extent that we don't introduce breaking changes. `convertToPresentation` is an overridden converter method that takes a `RatingScale` value and converts it into a `Double`.  In our `RatingScale` enum, we've set the `NO_RATING` to have the numeric value `null`. That's different to `ZERO` (a value of `0.0`). 

In other words, `null` is acceptable in this case as we handle this specifically. While we could change `value` to be `Optional`, I think that adds unnecessary noise with no real gain

Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/book/form/DoubleToRatingScaleConverter.java:38]: > It seems that just using get() here will produce some errors with BookFormTest.java.

Ignore me, you're right (`return RatingScale.toDouble(rating).orElse(null)`)

To test this, you can try adding a book to the read shelf. If you don't set a rating, it should show `No rating`, not `0/10`. This is why we need the `orElse(null)`

**Comment by knjk04** [comment]
@hieu-t-tran How are you getting on with this issue? If you need any help on this, feel free to reach out to us on Slack

If you need more time, that's fine, but please let us know.

**Comment by hieu-t-tran** [comment]
Thank you for checking in! I hopefully will be able to make an update by this Friday.

**Comment by knjk04** [comment]
How are you getting on with this?

**Comment by hieu-t-tran** [comment]
Hi there! Sorry for the delay. I am almost ready to update this pull request and just need to update some of the tests.

**Comment by knjk04** [comment]
@hieu-t-tran How are you getting on with this?

**Comment by hieu-t-tran** [comment]
I just made an update to the pull request and have a few questions:
* I am not sure how to update `addSeriesToTitle()` in `BookUtils.java`. Should we return an empty Optional when either of the 2 arguments is null?
* In `PredefinedShelfService.java`, should `findToReadShelf()` and the other 3 similar methods return a `PredefinedShelf` or an `Optional<PredefinedShelf>`?
* I notice that the `findById()` method in `PublisherService.java` calls a different function when compared to the `findById()` method in `AuthorService.java` and other service classes. Is this difference intentional?

**Comment by knjk04** [comment]
> I am not sure how to update addSeriesToTitle() in BookUtils.java. Should we return an empty Optional when either of the 2 arguments is null?

Yes, please. Also, the return type for `combineTitleAndSeries` would need to change to an Optional

> In PredefinedShelfService.java, should findToReadShelf() and the other 3 similar methods return a PredefinedShelf or an Optional<PredefinedShelf>?

Let's continue to return a `PredefinedShelf` for now (apologies for backtracking on this) for simplicity

> I notice that the findById() method in PublisherService.java calls a different function when compared to the findById() method in AuthorService.java and other service classes. Is this difference intentional?

Good spot! For the time being, we want to move to calling `findById()`, so feel free to change any occurrences of `getOne()` or something else if you see it.

You can see the accept answer on this [Stackoverflow question](https://stackoverflow.com/questions/24482117/when-use-getone-and-findone-methods-spring-data-jpa) for the difference

**Comment by knjk04** [comment]
@hieu-t-tran How are you getting on with this?

**Comment by hieu-t-tran** [comment]
Hi there! I think I covered everything with my last commits. Please let me know if there is any more work needed to be done. Thank you so much!

File: src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java [hunk 3]
```diff
@@ -88,28 +89,29 @@ void update(String chosenShelf, BookFilters bookFilters) {
...
         if (isAllBooksShelf(chosenShelf)) {
-            return null;
+            return Optional.empty();
         }
 
         if (PredefinedShelfUtils.isPredefinedShelf(chosenShelf)) {
-            return predefinedShelfService.getPredefinedShelfByNameAsString(chosenShelf);
+            return predefinedShelfService.getPredefinedShelfByNameAsString(chosenShelf)
+					.map(Shelf.class::cast);
         } else {
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/shelf/BookGrid.java:104]: I know this was already there, but could you please remove the `else`?

File: src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java [hunk 4]
```diff
@@ -261,7 +261,7 @@ void findSavedBook() {
         bookService.save(bookToSave);
 
         // then
-        assertThat(bookService.findById(bookToSave.getId())).isEqualTo(bookToSave);
+        assertThat(bookService.findById(bookToSave.getId()).get()).isEqualTo(bookToSave);
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/BookServiceTest.java:264]: Instead of using `get()` and `isEqualTo()`, could you please use `contains()`?

File: src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java [hunk 5]
```diff
@@ -85,14 +85,14 @@ void countOnlyReadBooksTowardsMostPagesStatistics() {
...
-        assertThat(pageStatistics.calculateAveragePageLength()).isNull();
+        assertThat(pageStatistics.calculateAveragePageLength()).isEmpty();
     }
 
     @Test
     void calculateAveragePageLengthCorrectly() {
         int averagePageLength =
                 StatisticTestUtils.getTotalNumberOfPages() / StatisticTestUtils.getNumberOfBooks();
-        assertThat(pageStatistics.calculateAveragePageLength()).isEqualTo(averagePageLength);
+        assertThat(pageStatistics.calculateAveragePageLength().get()).isEqualTo(averagePageLength);
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java:95]: `.contains()` can also be used here instead of `.get()` and `.isEqualTo()`

File: src/test/java/com/karankumar/bookproject/backend/statistics/util/StatisticTestUtils.java [hunk 6]
```diff
@@ -145,22 +146,16 @@ private static void saveBook(Book bookToSave) {
...
+        totalRating -= RatingScale.toDouble(ratingScale).orElse(0.0);
     }
     private static void updateThisYearRating(RatingScale ratingScale) {
-        Double rating = RatingScale.toDouble(ratingScale);
-        if (rating != null) {
-            thisYearRating += rating;
+        Optional<Double> rating = RatingScale.toDouble(ratingScale);
+        if (rating.isPresent()) {
+            thisYearRating += rating.get();
         }
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/util/StatisticTestUtils.java:159]: This can be simplified to `rating.ifPresent(x -> thisYearRating += x)`

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this. Please see my review comments

If you haven't already, I'd recommend installing the `SonarLint` plugin for your IDE. Some of my review suggestions were from SonarLint

File: src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java [hunk 7]
```diff
@@ -48,15 +48,15 @@ public CustomShelf createCustomShelf(String shelfName) {
         return new CustomShelf(shelfName, userService.getCurrentUser());
     }
 
-    public CustomShelf findById(Long id) {
-        return customShelfRepository.getOne(id);
+    public Optional<CustomShelf> findById(Long id) {
+        return Optional.ofNullable(customShelfRepository.getOne(id));
```
[the remark this answers is not in the document]
Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java:52]: If we change this to `findById()`, then it will fail the `deleteCustomShelf` test in `CustomShelfServiceTest.java`. It seems that the test method expects an exception to be thrown, which does not occur after the change.

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java [hunk 8]
```diff
@@ -89,22 +91,22 @@ public void save(@NonNull PredefinedShelf shelf) {
...
     }
 
     public PredefinedShelf findReadShelf() {
-        return findByPredefinedShelfNameAndLoggedInUser(READ);
+        return findByPredefinedShelfNameAndLoggedInUser(READ).get();
     }
 
     public PredefinedShelf findDidNotFinishShelf() {
-        return findByPredefinedShelfNameAndLoggedInUser(DID_NOT_FINISH);
+        return findByPredefinedShelfNameAndLoggedInUser(DID_NOT_FINISH).get();
```
[the remark this answers is not in the document]
Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java:106]: Just to confirm, do we just return null in the other case then? We should not use Optional for these methods for simplicity right?

File: src/main/java/com/karankumar/bookproject/backend/service/TagService.java [hunk 9]
```diff
@@ -33,8 +34,8 @@ public TagService(TagRepository tagRepository) {
         this.tagRepository = tagRepository;
     }
 
-    public Tag findById(Long id) {
-        return tagRepository.getOne(id);
+    public Optional<Tag> findById(Long id) {
+        return Optional.ofNullable(tagRepository.getOne(id));
```
[the remark this answers is not in the document]
Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/backend/service/TagService.java:38]: Similarly here, it will fail a test after the change.

File: src/main/java/com/karankumar/bookproject/ui/book/form/BookForm.java [hunk 10]
```diff
@@ -414,15 +412,15 @@ private Book populateBookBean() {
         PredefinedShelf predefinedShelf =
                 predefinedShelfService.findByPredefinedShelfNameAndLoggedInUser(
                         predefinedShelfField.getValue()
-                );
+                ).get();
```
[the remark this answers is not in the document]
Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/ui/book/form/BookForm.java:415]: Do you think we should use `orElse(null)` here?

**Comment by hieu-t-tran** [comment]
Thank you so much for your feedbacks! I have addressed most of the changes and left some questions for the ones I am not sure about. And thank you for recommending SonarLint - I have it installed now.

File: src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/service/CustomShelfService.java:52]: Could you please change the test so that it passes? You may only need to check whether the ID coming back is empty

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java]: Sorry, which other case?

File: src/main/java/com/karankumar/bookproject/backend/service/TagService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/service/TagService.java:38]: See previous comment

File: src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (hieu-t-tran) [reply src/main/java/com/karankumar/bookproject/backend/service/PredefinedShelfService.java]: The case in which `ifPresent()` returns false

File: src/main/java/com/karankumar/bookproject/ui/book/form/BookForm.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/book/form/BookForm.java:413]: Oops, I missed this one.

Perhaps it's best to return instead if this is null as it shouldn't be. Good question!

**Comment by knjk04** [comment]
How are you getting on with this?

**Comment by knjk04** [comment]
As we have not heard back from you in a while, despite following up. I will close this pull request. This is in line with our contributing guidelines
